workflows/testing/test_workflow/git.py

```diff
@@ -0,0 +1,20 @@
+import os
+import tempfile
+
+class GitRepository:
+    def __init__(self, url, ref):
+        self._url = url
+        self._dir = tempfile.TemporaryDirectory()
+        self.execute(f'git clone {url} {self._dir.name}')
+        self.execute(f'git checkout {ref}')
+        print(f'Checked out {url}@{ref} into {self._dir.name}')
+
+    def dir(self):
+        return self._dir.name
+
+    def execute(self, command):
```

Comment:
This doesn't fail when the command fails.

```python
import os
import tempfile
import subprocess

class GitRepository:
    def __init__(self, url, ref):
        self._url = url
        self._dir = tempfile.TemporaryDirectory()
        self.execute(f'git clone --branch {ref} --depth 1 {url} {self._dir.name}')
        print(f'Checked out {url}@{ref} into {self._dir.name}')

    def dir(self):
        return self._dir.name

    def execute(self, command):
        current = os.getcwd()
        print(f'Executing "{command}" in {self._dir.name}')
        os.chdir(self._dir.name)
        subprocess.check_call(command, shell = True)
        os.chdir(current)
```

Reply:
I didn't know about the `subprocess` library...that looks much better than `os.system`. Again, my python-fu is weak, so thanks for the pointer :)

Reply:
Beyond Python, make sure every method called fails loudly.

Reply:
Yes, of course. Remember that "the goal of this PR is to get feedback on the general approach, and to highlight the subtasks we should add to #122 for the next few days". Fault tolerance is not so important for prototype code.